### ISSUE TITLE
torch_geometric.utils.add_self_loops

### ISSUE THREAD

**Issue by downeykking** [pyg-team/pytorch_geometric#5014]
### 🛠 Proposed Refactor

When adding a self-loop to a given subgraph, the code has some problems logically:

`edge_index = torch.tensor([[0, 0, 1, 633, 1862],
                           [633, 1862, 1, 0, 0]])`
`edge_index = add_self_loops(edge_index, num_nodes=edge_index.size(1))`
`print(edge_index)`

it will return `tensor([[   0,    0,    1,  633, 1862,    0,    1,    2,    3,    4],
        [ 633, 1862,    1,    0,    0,    0,    1,    2,    3,    4]])`
tthe self loop of nodes 633, 1862 are not what we want.

### Suggest a potential alternative/fix

in function '[add_self_loops](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/utils/loop.html#add_self_loops)'
change code `loop_index = torch.arange(0, N, dtype=torch.long, device=edge_index.device)` to `loop_index = torch.unique(torch.cat([edge_index[0], edge_index[1]], dim=0))` may sovle the problem.

it will return `tensor([[   0,    0,    1,  633, 1862,    0,    1,  633, 1862],
        [ 633, 1862,    1,    0,    0,    0,    1,  633, 1862]])`

the same problem appears in '[add_remaining_self_loops](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/utils/loop.html#add_remaining_self_loops)' too.

**Comment by rusty1s** [comment]
I think this issue is resolved by just passing in `num_nodes=edge_index.max().item(). + 1`.

**Comment by downeykking** [comment]
> I think this issue is resolved by just passing in `num_nodes=edge_index.max().item(). + 1`.

In this way it will return `tensor([[   0,    0,    1,  ..., 1860, 1861, 1862],
        [ 633, 1862,    1,  ..., 1860, 1861, 1862]])`.
maybe more than we want

**Comment by rusty1s** [comment]
Yes, this is intended. Node indices should be always in between `0` and `num_nodes - 1`.
